### ISSUE TITLE
LDAP group shows zero members

### ISSUE THREAD

**Issue by gig13** [owncloud/core#20274]
### Steps to reproduce
1. Add some groups via the LDAP configuration to OC
2. Verify that the number of members are displayed in the count check
3. Verify that the number of members are displayed in the Users page
### Expected behaviour

When adding groups to OC via LDAP, these group members should display the number of members in the OC LDAP pages and User pages
### Actual behaviour

I was adding some groups and checking that the oc_ldap_group_members table refreshed and I noticed three groups that have many hundreds of members (and one with several thousand) that show with 0 members in the table. We do have paging setup with 500. I've tried smaller and larger. I've even tried 1001 to see if that'd cause a problem.

We have other groups with up to 800 users in the table.

One of the groups showing 0 members is owncloudusers, the group in our login filter. Logins are unaffected by this. I guess this table is used for group sharing?

Is this a known issue and related to AD's 1000 results per query issue? If so, it would appear paging is not working as it should.
### Server configuration

**Operating system**: RHEL 

**Web server:** Apache

**Database:** Galera Cluster

**PHP version:** 5.5.30

**ownCloud version:** 8.1.3

**Updated from an older ownCloud or fresh install:** Updated

**List of activated apps:** See s3 config report https://s3.owncloud.com/owncloud/index.php/apps/files?dir=%2FShared%2Fowncloud%2Fsupport%2Fgithub-issues%2Fcore%2F20274

**Are you using external storage, if yes which one:** No

**Are you using encryption:** NO

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory
#### LDAP configuration

See s3 config report

@MorrisJobke

**Comment by MorrisJobke** [comment]
@blizzz Could it be that this issue is solved by https://github.com/owncloud/core/pull/20191

@gig13 Is this AD or LDAP? Maybe https://github.com/owncloud/core/pull/19419 could also be a related issue. :see_no_evil:

**Comment by gig13** [comment]
@MorrisJobke 
AD

**Comment by blizzz** [comment]
@MorrisJobke no, #20191 fixes positive return message when testing loginnames in the wizard, #19419 prevents unnormalized DNs and possible duplication of groups.

The config looks okay, correct association attribute, group whitelisted in group filter, bases are alright.

Was it working before, or did  they check it only now? Are there any logs (covering background job run)? Are the users listed in the users page for the group in question?

>  I guess this table is used for group sharing?

This is required to throw hooks when users are leaving or added to groups. This was necessary for group sharing once, but the mechanism changed. While 3rd party apps may or may not depend on it, i am not aware that any supported app relies on this. @MorrisJobke correct me if i am wrong.

**Comment by MorrisJobke** [comment]
> @MorrisJobke correct me if i am wrong.

All correct.

**Comment by gig13** [comment]
@blizzz 
No groups are showing a count. Not even the local Admins group. No log messages (we're on warning on our production servers) when viewing the users page or selecting various groups.

I set the timestamp for the OCA\user_ldap\lib\Jobs job in the oc_jobs table to more than an hour ago and ran the cron script. No messages in the log. At warning level that only means that there aren't any uncaught exceptions or php errors. Because we only use these large groups on our production server, switching to debug logging would result in log spam and make it impossible to find relevant messages.

**Comment by gig13** [comment]
@blizzz 
The smallest new group is 1710. They have paging set in settings but it may not be working. All the other groups they use in owncloud are under 1000 and work fine.

**Comment by blizzz** [comment]
>  No groups are showing a count. Not even the local Admins group. No log messages (we're on warning on our production servers) when viewing the users page or selecting various groups.

Group counts on the Users page have been disabled for performance reasons when LDAP is used.

I will try to reproduce, but not sure I am able to get it done this week.

**Comment by davitol** [comment]
> Group counts on the Users page have been disabled for performance reasons when LDAP is used.
> 
Just FYI, if there are "normal" groups and LDAP conf is added, "normal" groups stop showing the group users count too.

**Comment by gig13** [comment]
@blizzz 
Any chance to test?

**Comment by blizzz** [comment]
I tried it now like this with OpenLDAP:

1. Assign ~1500 LDAP users to an LDAP group
2. Empty ``ldap_group_members`` database table
3. Run the background job

→ All group members have been written to the database. 

However I have a high soft and hard limit configured in LDAP server side. So, let's change this and repeat the steps above. To emulate AD, I set the soft limit to 1k, but keep the hard limit high (AD by default only has a soft limit of 1k). 

→ All group members have been written to the database. 

This now is with settings as expected. The group memberships are stored within the group's record, i.e. the number of members in the group does not matter. If at all, the number of groups.

Long story short, I could not reproduce the described issue. The only other hint that comes to my mind is to ensure that the cron job is executed as web user. However, with other privileges errors would show up, so I think that's alright. Btw, I ran it against current 8.1 from git, which is >= 8.1.4. 

Now I could only test it against OpenLDAP, because our AD test server vanished, and my external HD where I had a VM with one died. I already talked with @DeepDiver1975 this time about getting an AD up and running again.

**Comment by gig13** [comment]
@bboule @DeepDiver1975 
Can we provide access to the lab AD? Any other alternatives?

**Comment by blizzz** [comment]
I got one, just filling this up with test data.

**Comment by blizzz** [comment]
Seems I can reproduce it with AD. In my log appears ``Requested attribute member not found for $GROUPDN`` . Also on the users page these users are also not showing up, when selecting the group. Only in Everyone. Via command line, I am only getting 1500 results (1800 expected). These 1500 are a hard limit. Will look further into it.

**Comment by bboule** [comment]
@gig13 do we know what version of AD i being run?

**Comment by gig13** [comment]
@bboule 
2008 R2 -- main large domain
2012 -- secondary and considerably smaller

**Comment by bboule** [comment]
@blizzz does this help what would be the next steps here?

**Comment by blizzz** [comment]
The issue is following. AD has a limit an how many attribute values are being returned per LDAP entry, i.e. in this case "member" values for the group. If the number of members is smaller than the maximum value, everything is fine. However, if there are more members, two things happen:

1. Obviously, only as many results are returned as the maximum limit allows
2. The format of the result array changes :-/ 

Usually the result is an array having the attribute name as a key. This entry holds a sequential array of the values. On this edge case, this is also set, but empty. There is an additional key "[member;range=0-$MAXAMOUNT-1]" which contains the results. 

This means there are two possible ways to tackle this. 

1. On on the AD server side (non-ownCloud) by [setting the MaxValRange](http://www.fots.nl/set-the-maxpagesize/) to a number not smaller than the largest groups has members (and/or to match the recommended maximum group size).  This is what can be advised to the customer.
2. Add support on our side to deal with this. From outside it looks like a bugfix, but codewise it is an enhancement actually.

@bboule does it help?

**Comment by blizzz** [comment]
Right now, I am also not convinced that it is totally doable, as specifying the range does not seem to be standardized and there is no parameter visible in PHP's LDAP methods on first glances.

**Comment by blizzz** [comment]
For the record, resources on this matter:
* https://support.microsoft.com/en-us/kb/2009267
* https://msdn.microsoft.com/en-us/library/cc223242.aspx
* https://www.ietf.org/rfc/rfc2251.txt (section 4.5.1)

**Comment by blizzz** [comment]
OK, this caught me and I did a "proof of concept" of getting the ranged support implemented, and here it is: https://github.com/owncloud/core/pull/21525

Since it changes the heavily used readAttribute() method, I would prefer to not backport it, and only introduce it as feature in 9.0.

**Comment by blizzz** [comment]
Suggest to move to  9.1, see https://github.com/owncloud/core/pull/21525#issuecomment-183502741

**Comment by cmonteroluque** [comment]
@bboule @MorrisJobke moving to 9.1 unless there's an objection

**Comment by MorrisJobke** [comment]
> @bboule @MorrisJobke moving to 9.1 unless there's an objection

@gig13 Is this okay for you?

**Comment by gig13** [comment]
@MorrisJobke Yes

**Comment by cdamken** [comment]
@PVince81 @jvillafanez Will the this on 9.1.1?

**Comment by jvillafanez** [comment]
#21525 seems to be the one that solves the issue, and it isn't merged yet, so I guess this will be in 9.2 unless we backport the solution.

**Comment by PVince81** [comment]
Fix is in 10.0 beta
